core/dbt/clients/yaml_helper.py

```diff
@@ -61,3 +65,42 @@ def load_yaml_text(contents, path=None):
             error = str(e)
 
         raise dbt.exceptions.DbtValidationError(error)
+
+
+def split_yaml_frontmatter(content: str) -> Tuple[Optional[str], str]:
+    """Splits `content` into raw YAML frontmatter (as a raw string) and everything else proceeding.
+
+    Frontmatter is defined as a block of YAML appearing between two `---` tokens in an otherwise non-YAML document.
+    The frontmatter must be placed at the beginning of the file: anything other than whitespace preceding the first `---`
+    will cause the frontmatter block to be ignored, with a warning.
+    """
+    parts = FRONTMATTER_DELIMITER.split(content, 2)
+    if len(parts) != 3:
+        # Zero or one `---` token, so return the original string
+        return None, content
+    elif NON_WHITESPACE.search(parts[0]) is not None:
+        # No frontmatter section or non-whitespace preceding the first `---`, so skip frontmatter
+        return None, content
+
+    frontmatter_content, after_footer = parts[1:]
+    return frontmatter_content, after_footer
+
+
+def parse_yaml_frontmatter(frontmatter_content: str, original_content: str):
+    try:
+        parsed_yaml = safe_load(frontmatter_content)
+    except (yaml.scanner.ScannerError, yaml.YAMLError) as e:
+        if hasattr(e, "problem_mark"):
+            error = contextualized_yaml_error(original_content, e)
+        else:
+            error = str(e)
+        error = f"Error parsing YAML frontmatter:  {error}"
+        raise dbt.exceptions.DbtValidationError(error)
+
+    return parsed_yaml
+
+
+def has_yaml_frontmatter(content: str) -> bool:
+    """Check first line for yaml frontmatter"""
+
+    return content.startswith(FRONTMATTER_CHECK[0]) or content.startswith(FRONTMATTER_CHECK[1])
```

Comment:
I think `content` could start with whitespace (including newlines) and this method would return `False`, while `split_yaml_frontmatter` would skip over leading whitespace.

Perhaps this would be better?

```suggestion
    stripped = content.lstrip()
    return stripped.startswith(FRONTMATTER_CHECK[0]) or stripped.startswith(FRONTMATTER_CHECK[1])
```

Reply:
Content should be stripped already, I think, when it's loaded in read_files. Did you actually see an error? Wouldn't hurt to double up on that though.

Reply:
Nope! I honestly just drove by and looked over your approach. Cool that it's already stripped. It might be worth a test case?